### ISSUE TITLE
[typescript-react-apollo] Create options for typed QueryResult, MutationResult and MutationOptions

### ISSUE THREAD

**Issue by eddeee888** [dotansimha/graphql-code-generator#2144]
In my projects, I'm using `typescript-react-apollo` plugin and I find myself having to type things like `QueryResult`, `MutationResult` and `MutationOptions` quite often as I use these types in other components quite often. For example:

```ts
// For a Query
export type ViewUserQueryResult = ReactApollo.QueryResult<ViewUserQuery, ViewUserQueryVariables>
```

```ts
// For a Mutation
export type LoginMutationResult = ReactApollo.MutationResult<LoginMutation>
export type LoginMutationOptions = ReactApollo.MutationOptions<LoginMutation, LoginMutationVariables>
```

There's already a config for `withMutationFn` as seen [here](https://github.com/dotansimha/graphql-code-generator/blob/53ecdef9b8789f5a70285c2e917bec620a24e66c/packages/plugins/typescript/react-apollo/src/visitor.ts#L74-L79). Is there a plan to support other types? 

At the moment, I have forked from master and created options called:
- `withResultType`: boolean, will generate `*QueryResult` and `*MutaionResult`
- `withMutationOptionsType`: boolean, will generate `*MutationOptions`

I'm happy to create a PR to add these options if that's a feature we want to have?

**Comment by dotansimha** [comment]
@eddeee888 a PR would be great. We have this for hooks (https://github.com/dotansimha/graphql-code-generator/pull/2133), but not for component/hoc.
Thanks!

**Comment by 18601673727** [comment]
Hey sorry for the interrupt, but maybe I'm having troubles with this line:
https://github.com/dotansimha/graphql-code-generator/blob/474a951c5f46daa01517eec8ce6ca6cb8f1b5a17/packages/plugins/typescript/react-apollo/src/visitor.ts#L174

which maybe generates following code:
```ts
export type SaveTagsMutationOptions = ReactApollo.MutationOptions<
  SaveTagsMutation,
  SaveTagsMutationVariables
>;
```
IDE will complain about error like below:
```
Namespace '"/Users/cool-user/awesome-project/node_modules/@apollo/react-common/lib/index"' has no exported member 'MutationOptions'.ts(2694)
```
Changed `ReactApollo` to `ReactApolloHooks`, seems resolved.

Let me know if this helped, thank you all!

**Comment by eddeee888** [comment]
Hi @18601673727 !

This is supposed to be used for the non-hook version i.e. `react-apollo`. I'm not using `react-apollo-hooks` so I haven't accounted for that :) Maybe someone who has used it before can create a PR for the hook version? 

Also, it's worth to double-check if the type name is different between official beta `@apollo/react-hooks` vs `react-apollo-hooks` type name

**Comment by 18601673727** [comment]
Hi @eddeee888 !

Got it, this is actually from `@apollo/react-hooks` integrated build, you could find my setup [here](https://github.com/dotansimha/graphql-code-generator/issues/2288). Your fantastic PR already covers a lot, maybe just a few more steps will support both! =)

**Comment by dotansimha** [comment]
Available in 1.5.0 🎉
